[JIT][read path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54746 [JIT][write path] Make NoneType annotation_str emit `NoneType` instead of `None`
* #54745 [JIT][write path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type
* #54642 [JIT][read path] Make NoneType annotation_str emit `NoneType` instead of `None`
* **#54641 [JIT][read path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type**

Previously, `TupleType` would emit the string `Tuple[]` for a `TupleType` with 0 contained elements. This is not valid Python syntax and is a divergence from what the `typing` module does: `typing` defines `Tuple[()]` as a 0-element Tuple type (see https://docs.python.org/3/library/typing.html#typing.Tuple). This change updates `TupleType::annotation_str` to conform to that as well as updated the script type parser to accept this syntax. It should be BC-compatible since the `Tuple[]` syntax is still accepted on the read path.

Differential Revision: [D27314173](https://our.internmc.facebook.com/intern/diff/D27314173)

Commit:
[JIT] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type"


Previously, `TupleType` would emit the string `Tuple[]` for a `TupleType` with 0 contained elements. This is not valid Python syntax and is a divergence from what the `typing` module does: `typing` defines `Tuple[()]` as a 0-element Tuple type (see https://docs.python.org/3/library/typing.html#typing.Tuple). This change updates `TupleType::annotation_str` to conform to that as well as updated the script type parser to accept this syntax. It should be BC-compatible since the `Tuple[]` syntax is still accepted on the read path.

Differential Revision: [D27314173](https://our.internmc.facebook.com/intern/diff/D27314173)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type"


Previously, `TupleType` would emit the string `Tuple[]` for a `TupleType` with 0 contained elements. This is not valid Python syntax and is a divergence from what the `typing` module does: `typing` defines `Tuple[()]` as a 0-element Tuple type (see https://docs.python.org/3/library/typing.html#typing.Tuple). This change updates `TupleType::annotation_str` to conform to that as well as updated the script type parser to accept this syntax. It should be BC-compatible since the `Tuple[]` syntax is still accepted on the read path.

Differential Revision: [D27314173](https://our.internmc.facebook.com/intern/diff/D27314173)

[ghstack-poisoned]

SplitInfinity commented on test/test_jit.py:
```diff
@@ -11547,6 +11548,12 @@ def range_non_float():
+    def test_empty_tuple_str(self):
+        empty_tuple_type = torch._C.TupleType([])
+        g = {'Tuple' : typing.Tuple}
+        python_type = eval(empty_tuple_type.annotation_str, g)
```
🤯

SplitInfinity commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -27,6 +27,19 @@ TypePtr ScriptTypeParser::subscriptToType(
+      // here. See https://docs.python.org/3/library/typing.html#typing.Tuple
+      auto tup_literal = TupleLiteral(subscript.subscript_exprs()[0]);
+      if (tup_literal.inputs().size() > 0) {
+        throw ErrorReport(tup_literal.range())
+            << "Tuple literal in Tuple type annotation must not "
+            << "have any elements!";
```
Can you add a test for this error message?

SplitInfinity reviewed (commented):
cc: @Lilyjjo 

I remember you used this quirk (or a similar one) when you enabled hooks. Do you think this will break anything?

Forward compatibility concern from #54642 is applicable here too. The usual solution is to land the read path changes first, and then the write path changes after some time.

Commit:
Update on "[JIT] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type"


Previously, `TupleType` would emit the string `Tuple[]` for a `TupleType` with 0 contained elements. This is not valid Python syntax and is a divergence from what the `typing` module does: `typing` defines `Tuple[()]` as a 0-element Tuple type (see https://docs.python.org/3/library/typing.html#typing.Tuple). This change updates `TupleType::annotation_str` to conform to that as well as updated the script type parser to accept this syntax. It should be BC-compatible since the `Tuple[]` syntax is still accepted on the read path.

Differential Revision: [D27314173](https://our.internmc.facebook.com/intern/diff/D27314173)

[ghstack-poisoned]

Commit:
Update on "[JIT][read path] Fix TupleType.annotation_str to conform to `typing` module syntax for empty tuple type"


Previously, `TupleType` would emit the string `Tuple[]` for a `TupleType` with 0 contained elements. This is not valid Python syntax and is a divergence from what the `typing` module does: `typing` defines `Tuple[()]` as a 0-element Tuple type (see https://docs.python.org/3/library/typing.html#typing.Tuple). This change updates `TupleType::annotation_str` to conform to that as well as updated the script type parser to accept this syntax. It should be BC-compatible since the `Tuple[]` syntax is still accepted on the read path.

Differential Revision: [D27314173](https://our.internmc.facebook.com/intern/diff/D27314173)

[ghstack-poisoned]
